src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Helpers.cs

```diff
@@ -0,0 +1,138 @@
+        {
+            Array.Resize(ref a, size);
+#if DEBUG
+            a = a.CloneByteArray();
+#endif
+            return a;
+        }
+
+        public static CmsRecipientCollection DeepCopy(this CmsRecipientCollection recipients)
+        {
+            CmsRecipientCollection recipientsCopy = new CmsRecipientCollection();
+            foreach (CmsRecipient recipient in recipients)
+            {
+                X509Certificate2 originalCert = recipient.Certificate;
+                X509Certificate2 certCopy = new X509Certificate2(originalCert.Handle);
+                CmsRecipient recipientCopy = new CmsRecipient(recipient.RecipientIdentifierType, certCopy);
+                recipientsCopy.Add(recipientCopy);
+                GC.KeepAlive(originalCert);
+            }
+            return recipientsCopy;
+        }
+
+        public static X509Certificate2Collection GetStoreCertificates(StoreName storeName, StoreLocation storeLocation, bool openExistingOnly)
+        {
+            using (X509Store store = new X509Store())
+            {
+                OpenFlags flags = OpenFlags.ReadOnly | OpenFlags.IncludeArchived;
+                if (openExistingOnly)
+                    flags |= OpenFlags.OpenExistingOnly;
+
+                store.Open(flags);
+                X509Certificate2Collection certificates = store.Certificates;
+                return certificates;
+            }
+        }
+
+        public static X509Certificate2 TryFindMatchingCertificate(this X509Certificate2Collection certs, SubjectIdentifier recipientIdentifier)
+        {
+            SubjectIdentifierType recipientIdentifierType = recipientIdentifier.Type;
+            X509Certificate2Collection matches;
+            switch (recipientIdentifierType)
+            {
+                case SubjectIdentifierType.IssuerAndSerialNumber:
+                    {
+                        X509IssuerSerial issuerSerial = (X509IssuerSerial)(recipientIdentifier.Value);
+                        matches = certs.Find(X509FindType.FindByIssuerDistinguishedName, issuerSerial.IssuerName, validOnly: false);
+                        matches = matches.Find(X509FindType.FindBySerialNumber, issuerSerial.SerialNumber, validOnly: false);
+                        break;
+                    }
+
+                case SubjectIdentifierType.SubjectKeyIdentifier:
+                    {
+                        string ski = (string)(recipientIdentifier.Value);
+                        matches = certs.Find(X509FindType.FindBySubjectKeyIdentifier, ski, validOnly: false);
+                        break;
+                    }
+
+                default:
+                    // RecipientInfo's can only be created by this package so if this an invalid type, it's the package's fault.
+                    Debug.Assert(false, $"Invalid recipientIdentifier type: {recipientIdentifierType}");
+                    throw new CryptographicException();
```

Comment:
Will change Debug.Assert() to Debug.Fail()

I don’t see the compat issue here – only the PKCS assembly can construct SubjectIdentifier objects so if there’s an illegal value there, is an internal PKCS bug (or a Reflection invasion, in which case, I don’t care about compat.)

So I’m intentionally doing a different thing here. There are few other cases where the original ECMS code did “api-style” validation for non-public constructors – I changed those to “Assert and throw something to keep the compiler happy” as well.


From: Jeremy Barton [mailto:notifications@github.com]
Sent: Tuesday, April 26, 2016 10:31 AM
To: dotnet/corefx <corefx@noreply.github.com>
Cc: Atsushi Kanamori <atsushik@microsoft.com>; Author <author@noreply.github.com>
Subject: Re: [dotnet/corefx] Add Support for EnvelopedCms (#8083)


In src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Helpers.cs<https://github.com/dotnet/corefx/pull/8083#discussion_r61128618>:

> +                        matches = certs.Find(X509FindType.FindByIssuerDistinguishedName, issuerSerial.IssuerName, validOnly: false);

> +                        matches = matches.Find(X509FindType.FindBySerialNumber, issuerSerial.SerialNumber, validOnly: false);

> +                        break;

> +                    }

> +

> +                case SubjectIdentifierType.SubjectKeyIdentifier:

> +                    {

> +                        string ski = (string)(recipientIdentifier.Value);

> +                        matches = certs.Find(X509FindType.FindBySubjectKeyIdentifier, ski, validOnly: false);

> +                        break;

> +                    }

> +

> +                default:

> +                    // RecipientInfo's can only be created by this package so if this an invalid type, it's the package's fault.

> +                    Debug.Assert(false, $"Invalid recipientIdentifier type: {recipientIdentifierType}");

> +                    throw new CryptographicException();

Compat: Desktop seems to return null in this case.
Coding: Debug.Fail instead of Assert(false)?

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/dotnet/corefx/pull/8083/files/dc133683887dcca84a90a93ff47728ab42d7eba1#r61128618>